### PR TITLE
(MODULES-5264) allow UNC PhysicalPath in iis_virtual_directory

### PR DESCRIPTION
**Pull request by tkishel**
Adds the 'Path` class to allow reuse of an accurate path validation regex  since paths are attributes of multiple types. Adds methods to test whether a path is local or remote, and if a path is local, whether the path exists.

Refactors the type to use the Path class for validation. Refactors the provider to use path methods for validation, allows the provider to create and update virtual directories with UNC paths. Adds unit tests.

### CODE REVIEW

File: lib/puppet/provider/iis_common.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+def is_unc_path(path)
+  return (path =~ /^\\\\[^\\]+\\[^\\]+/)
+end
+
+def verify_physicalpath
+  if @resource[:physicalpath].nil? or @resource[:physicalpath].empty?
+    fail("physicalpath is a required parameter")
+  end
+  if is_local_path(@resource[:physicalpath])
+    if ! File.exists?(@resource[:physicalpath])
```
Comment (Iristyle) [inline lib/puppet/provider/iis_common.rb:14]: No existence checks for UNC paths 👍

**Comment by tkishel** [comment]
This PR follows the comments in PE-22667.

File: lib/puppet/provider/iis_virtual_directory/webadministration.rb [hunk 2]
```diff
@@ -40,25 +41,32 @@ def create
...
-      fail "physicalpath doesn't exist: #{@resource[:physicalpath]}"
-    end
+    verify_physicalpath
 
     cmd = []
 
-    cmd << "Set-ItemProperty -Path 'IIS:\\Sites\\#{@resource[:sitename]}\\#{@resource[:name]}' -Name 'site' -Value '#{@resource[:sitename]}';" if @resource[:sitename]
-    cmd << "Set-ItemProperty -Path 'IIS:\\Sites\\#{@resource[:sitename]}\\#{@resource[:name]}' -Name 'physicalpath' -Value '#{@resource[:physicalpath]}';" if @resource[:physicalpath]
-    cmd << "Set-ItemProperty -Path 'IIS:\\Sites\\#{@resource[:sitename]}\\#{@resource[:name]}' -Name 'application' -Value '#{@resource[:application]}';" if @resource[:application]
+    # Set-ItemProperty fails when setting site or application:
```
Comment (jpogran) [inline lib/puppet/provider/iis_virtual_directory/webadministration.rb]: Does this mean the name of the site or the application name cannot be changed after creation? If so, we need change them from parameters to properties

File: lib/puppet/provider/iis_virtual_directory/webadministration.rb [hunk 3]
```diff
@@ -40,25 +41,32 @@ def create
   end
 
   def update
+    if (@resource[:ensure] == :absent)
```
Comment (jpogran) [inline lib/puppet/provider/iis_virtual_directory/webadministration.rb]: This still seems like it should be a separate ticket/fix as it doesn't haven anything to do with path validation, its fixing a core part of the provider that shouldn't be firing. I think we can let this stay in so long a new bug ticket is added for this to investigate why this provider is acting like you report.

File: lib/puppet_x/puppetlabs/iis/property/path.rb [hunk 4]
```diff
@@ -0,0 +1,17 @@
+# (c:\directory or c:/directory) or \\server\share
```
Comment (jpogran) [inline lib/puppet_x/puppetlabs/iis/property/path.rb]: Move the comment closer to the regex, or remove from top of file

File: spec/unit/puppet/type/iis_virtual_directory_spec.rb [hunk 5]
```diff
@@ -49,24 +34,75 @@
     expect(params + [:provider]).to include(*type_class.parameters)
   end
 
-  it 'should default ensure to present' do
-    pool = type_class.new(
-      name: 'foo',
-    )
-    expect(pool[:ensure]).to eq(:present)
+  let(:resource) { described_class.new(:name => "iis_virtual_directory") }
```
Comment (jpogran) [inline spec/unit/puppet/type/iis_virtual_directory_spec.rb:37]: Great job cleaning up these tests

**Review by jpogran — CHANGES_REQUESTED** [review]
Minimal work needed. Great job otherwise

File: lib/puppet/provider/iis_virtual_directory/webadministration.rb (hunk 3 above)
Comment (tkishel) [reply lib/puppet/provider/iis_virtual_directory/webadministration.rb]: MODULES-6062 documents the update after destroy issue. In response to the request in an earlier version of this PR, I added `ErrorAction Stop` to these PowerShell commands. They will throw an error if called on a deleted virtual directory.

File: lib/puppet/provider/iis_virtual_directory/webadministration.rb (hunk 2 above)
Comment (tkishel) [reply lib/puppet/provider/iis_virtual_directory/webadministration.rb]: Yes, site and application cannot be modified with Set-ItemProperty. MODULES-6061 documents this issue. I decided to limit the scope of this PR by commenting out those commands so they do not interfere with the command that can modify physicalpath.

File: lib/puppet_x/puppetlabs/iis/property/path.rb (hunk 4 above)
Comment (tkishel) [reply lib/puppet_x/puppetlabs/iis/property/path.rb]: Deleted.

**Review by jpogran — APPROVED** [review]

**Comment by jpogran** [comment]
Adhoc passed

**Pull request merged**